### ISSUE TITLE
Change datetime type for samples read_meter_data_from_csv

### ISSUE THREAD

**Issue by limejoeg** [openeemeter/eemeter#371]
When loading the default sample data, .tz_localize('UTC') requires a different datetime type than is being passed. The parser is unable to standardize dates and is unable to convert 'start' to a datetime index. 

I fixed this bug by explicitly converting the datetime column. 

SOLUTION:
def meter_data_from_csv(
    filepath_or_buffer,
    tz=None,
    start_col="start",
    value_col="value",
    gzipped=False,
    freq=None,
    **kwargs
):
    """ Load meter data from a CSV file.
    Default format::
        start,value
        2017-01-01T00:00:00+00:00,0.31
        2017-01-02T00:00:00+00:00,0.4
        2017-01-03T00:00:00+00:00,0.58
    Parameters
    ----------
    filepath_or_buffer : :any:`str` or file-handle
        File path or object.
    tz : :any:`str`, optional
        E.g., ``'UTC'`` or ``'US/Pacific'``
    start_col : :any:`str`, optional, default ``'start'``
        Date period start column.
    value_col : :any:`str`, optional, default ``'value'``
        Value column, can be in any unit.
    gzipped : :any:`bool`, optional
        Whether file is gzipped.
    freq : :any:`str`, optional
        If given, apply frequency to data using :any:`pandas.DataFrame.resample`.
    **kwargs
        Extra keyword arguments to pass to :any:`pandas.read_csv`, such as
        ``sep='|'``.
    """

    read_csv_kwargs = {
        "usecols": [start_col, value_col],
        "dtype": {value_col: np.float64},
        "parse_dates": [start_col],
        "index_col": start_col,
    }

    if gzipped:
        read_csv_kwargs.update({"compression": "gzip"})

    # allow passing extra kwargs
    read_csv_kwargs.update(kwargs)

    df = pd.read_csv(filepath_or_buffer, **read_csv_kwargs)
    **df.index = pd.to_datetime(df.index, utc=True)**
    if tz is not None:
        df = df.tz_convert(tz)

    if freq == "hourly":
        df = df.resample("H").sum()
    elif freq == "daily":
        df = df.resample("D").sum()

    return df

ERROR:

TypeError                                 Traceback (most recent call last)
<ipython-input-8-c46e5679ce40> in <module>
      1 #Daily Billing for Caltrack
      2 meter_data, temperature_data, sample_metadata = (
----> 3     eemeter.load_sample("il-electricity-cdd-hdd-daily")
      4 )
      5 

~/anaconda3/envs/eenv/lib/python3.7/site-packages/eemeter/samples/load.py in load_sample(sample)
     80     meter_data_filename = metadata["meter_data_filename"]
     81     with resource_stream("eemeter.samples", meter_data_filename) as f:
---> 82         meter_data = meter_data_from_csv(f, gzipped=True, freq=freq)
     83 
     84     temperature_filename = metadata["temperature_filename"]

~/anaconda3/envs/eenv/lib/python3.7/site-packages/eemeter/io.py in meter_data_from_csv(filepath_or_buffer, tz, start_col, value_col, gzipped, freq, **kwargs)
     81     read_csv_kwargs.update(kwargs)
     82 
---> 83     df = pd.read_csv(filepath_or_buffer, **read_csv_kwargs).tz_localize("UTC")
     84     if tz is not None:
     85         df = df.tz_convert(tz)

~/anaconda3/envs/eenv/lib/python3.7/site-packages/pandas/core/generic.py in tz_localize(self, tz, axis, level, copy, ambiguous, nonexistent)
   9865             if level not in (None, 0, ax.name):
   9866                 raise ValueError("The level {0} is not valid".format(level))
-> 9867             ax = _tz_localize(ax, tz, ambiguous, nonexistent)
   9868 
   9869         result = self._constructor(self._data, copy=copy)

~/anaconda3/envs/eenv/lib/python3.7/site-packages/pandas/core/generic.py in _tz_localize(ax, tz, ambiguous, nonexistent)
   9848                     ax_name = self._get_axis_name(axis)
   9849                     raise TypeError(
-> 9850                         "%s is not a valid DatetimeIndex or " "PeriodIndex" % ax_name
   9851                     )
   9852                 else:

TypeError: index is not a valid DatetimeIndex or PeriodIndex

**Comment by philngo** [comment]
@limejoeg Thank you for your bug fix. A couple of things that would help me include this in the next version:

1. Would you consider making a fork, putting your changes into a branch, and creating a [pull request](https://github.com/openeemeter/eemeter/compare) ([how to](https://help.github.com/en/articles/creating-a-pull-request))? This will help us run automated tests on your branch so that we can see if your change is backwards compatible.

2. Would you share a sample of data (or at least the datetime format) you are reading in so that case can be added to the [unit tests](https://github.com/openeemeter/eemeter/blob/cac2d50566ab32cbbbe697bbe905d0940ec7baa9/tests/test_io.py#L91)?

**Comment by limejoeg** [comment]
Hey Phil,

If I have time I will try to do that. It’s really just a one line fix that is happening bc of Pandas updates (I think). I’m getting the error when I go through the tutorial. The error happens for any of the sample files. I think the index has to be explicitly converted to datetime before tz_convert can be used.
Calling:

meter_data, temperature_data, sample_metadata = (
    eemeter.load_sample("il-electricity-cdd-hdd-daily")
)

was what threw the error.

And here is the fix if you’d like to quickly put it in.

Line 83 of io.py

df = pd.read_csv(filepath_or_buffer, **read_csv_kwargs).tz_localize("UTC")
#add in
df.index = pd.to_datetime(df.index, utc=True)

Hope this helps. At some point I anticipate contributing more, but I will not have time to do anything in the near future. Please let me know if you have any questions, or think that my assumptions are off.

Joe Glass
Data Science
Lime Energy
510.919.8200

**Comment by limejoeg** [comment]
This bug is for the temperature_data_from_csv as well.

From: Joseph Glass-Katz <Joe.Glass@Lime-Energy.com>
Date: Tuesday, October 1, 2019 at 11:26 AM
To: openeemeter/eemeter <reply@reply.github.com>
Subject: Re: [openeemeter/eemeter] Change datetime type for samples read_meter_data_from_csv (#371)

Hey Phil,

If I have time I will try to do that. It’s really just a one line fix that is happening bc of Pandas updates (I think). I’m getting the error when I go through the tutorial. The error happens for any of the sample files. I think the index has to be explicitly converted to datetime before tz_convert can be used.
Calling:

meter_data, temperature_data, sample_metadata = (
    eemeter.load_sample("il-electricity-cdd-hdd-daily")
)

was what threw the error.

And here is the fix if you’d like to quickly put it in.

Line 83 of io.py

df = pd.read_csv(filepath_or_buffer, **read_csv_kwargs).tz_localize("UTC")
#add in
df.index = pd.to_datetime(df.index, utc=True)

Hope this helps. At some point I anticipate contributing more, but I will not have time to do anything in the near future. Please let me know if you have any questions, or think that my assumptions are off.

Joe Glass
Data Science
Lime Energy
510.919.8200

**Comment by philngo** [comment]
@limejoeg Ah, I see. Yeah, this will help us support the newest pandas versions (0.24.x), we've been stuck on 0.23.x for a while. If I get to it first I'll make a PR and tag you for review.

**Comment by opentaps** [comment]
FYI we were just trying eemeter today and got a similar bug using 2.7.5 from pip in a python3 venv with pandas 0.25.1 or 0.24.2:
```
Python 3.7.2 (default, Feb 20 2019, 14:37:21)
[Clang 10.0.0 (clang-1000.11.45.5)] on darwin
>>> import eemeter
>>> eemeter.samples()
['il-electricity-cdd-hdd-billing_bimonthly', 'il-electricity-cdd-hdd-billing_monthly', 'il-electricity-cdd-hdd-daily', 'il-electricity-cdd-hdd-hourly', 'il-electricity-cdd-only-billing_bimonthly', 'il-electricity-cdd-only-billing_monthly', 'il-electricity-cdd-only-daily', 'il-electricity-cdd-only-hourly', 'il-gas-hdd-only-billing_bimonthly', 'il-gas-hdd-only-billing_monthly', 'il-gas-hdd-only-daily', 'il-gas-hdd-only-hourly', 'il-gas-intercept-only-billing_bimonthly', 'il-gas-intercept-only-billing_monthly', 'il-gas-intercept-only-daily', 'il-gas-intercept-only-hourly']
>>> meter_data, temperature_data, metadata = eemeter.load_sample('il-electricity-cdd-hdd-daily')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "venv/lib/python3.7/site-packages/eemeter/samples/load.py", line 82, in load_sample
    meter_data = meter_data_from_csv(f, gzipped=True, freq=freq)
  File "venv/lib/python3.7/site-packages/eemeter/io.py", line 83, in meter_data_from_csv
    df = pd.read_csv(filepath_or_buffer, **read_csv_kwargs).tz_localize("UTC")
  File "venv/lib/python3.7/site-packages/pandas/core/generic.py", line 9867, in tz_localize
    ax = _tz_localize(ax, tz, ambiguous, nonexistent)
  File "venv/lib/python3.7/site-packages/pandas/core/generic.py", line 9850, in _tz_localize
    "%s is not a valid DatetimeIndex or " "PeriodIndex" % ax_name
TypeError: index is not a valid DatetimeIndex or PeriodIndex
```
It only works when downgrading to pandas 0.23.4.

Replacing in io.py line 83 and line 149, the two instances of
```
    df = pd.read_csv(filepath_or_buffer, **read_csv_kwargs).tz_localize(tz)
```
with
```
    df = pd.read_csv(filepath_or_buffer, **read_csv_kwargs)
    df.index = pd.to_datetime(df.index, utc=True)
``` 

Did fix the exception.

**Comment by philngo** [comment]
@limejoeg @opentaps I went ahead and implemented the fix in #372. I needed to make a few modifications to retain compatibility with pandas<0.24.0, but it looks like we should be good now with both the recent and older pandas versions. I've released the fix in eemeter=2.7.6.

I hope that both of you will consider helping our developer community by filling out our [first-time issue/PR contributor survey](https://forms.gle/egrBUVm81g7GVTk28).
